Task 10

YaroslavKomarov opened the pull request.

YaroslavKomarov committed:
Added .gitignore.

YaroslavKomarov committed:
practise start

YaroslavKomarov committed:
Removed additional methods;
Fixed the .gitignore file.

YaroslavKomarov committed:
Deleted .vcxproj file;
Deleted .sln file;
Deleted .vcxproj.filters file.

YaroslavKomarov committed:
Removed the additional GetChar method

YaroslavKomarov committed:
Added class MTreeNode;
Improved the hasConnection method in Maze.cpp;
A program has been written to display a new maze and tree.

YaroslavKomarov committed:
В заголовочных вайлах теперь только объявления пременных и функций.

YaroslavKomarov committed:
Новый метод searchNode добавлен в класс MTreeNode.

YaroslavKomarov committed:
В Practise_18.11.20.cpp добавлены методы для реализации task 10.

Toxin65 commented:
Определитесь, пожалуйста, что из этого мне смотреть.
Заявлено было 8,9. Вижу два ПР, и только в одном что-то про Таск 10...

YaroslavKomarov commented:
Я, честно говоря, затрудняюсь ответить. Я проходил ревью у Елены Александровны и она мне сказала, внести последнюю правку, после чего она могла бы принять у меня task 8 и task 9. Однако к тому моменту я уже выполнили и task 10, который по просьбе Елены Александровны поместил в отдельный pull

Toxin65 commented:
Ну, я при просмотре 10го накидаю замечаний по всем трём, чует моё сердце...
Так что может разницы то и нет.
Тем не менее, мне нужны чёткие указания. Что мне смотреть и в какую графу потом писать, соответственно.

YaroslavKomarov commented:
Елена Александровна мне сказала, что Вы будете проверять у меня 10 задание.

YaroslavKomarov commented:
Также Елена Александровна сказала, что сможет зачесть мне 8 и 9 задание, если Вы зачтете мне 10)

Toxin65 commented:
Ок, гляжу.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MCell.h:
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+class MCell {
+private:
```
Сначала public.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MCell.cpp:
```diff
@@ -0,0 +1,10 @@
+
+MCell::MCell() {
+	m_down = false;
+	m_right = false;
```
Современный стандарт позволяет инициализировать при объявлении полей.
Так и следует поступать, чтобы ничего не забыть.
Инициализировать следует все поля.
А конструктор достаточно  = default

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MCell.cpp:
```diff
@@ -0,0 +1,10 @@
+	m_right = false;
+}
+
+bool MCell::down() const { return m_down; }
```
Строго говоря, однострочные принято писать в заголовочных файлах прямо при объявлении.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.h:
```diff
@@ -0,0 +1,26 @@
+#pragma once
+#include <vector>
+using namespace std;
```
using namespace в заголовочных файлах не пишут.
Ибо это повлияет на контекст тех, кто их подключает, а не должно!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.h:
```diff
@@ -0,0 +1,26 @@
+private:
+	int m_i, m_j;
+	int m_distance;
+	MTreeNode* m_parent;
```
Все поля инициализируем обязательно при объявлении!!!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.h:
```diff
@@ -0,0 +1,26 @@
+	int childCount() const;
+	bool addChild(int i, int j);
+	const MTreeNode* parent() const;
+	MTreeNode* hasChild(int i, int j);
```
const

YaroslavKomarov replied:
Прошу прощения, не совсем понял что именно сделать константным? В тз task9.txt сигнатура всех этих методов такая же как и у меня

YaroslavKomarov replied:
Понял

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.h:
```diff
@@ -0,0 +1,26 @@
+	MTreeNode* hasChild(int i, int j);
+	const MTreeNode* child(int i) const;
+	static MTreeNode* begintTree(int i, int j);
+	static MTreeNode* searchNode(const MTreeNode& tree, const int i, const int j);
```
Бред. Зачем его делать static-ом и сразу же передавать ссылку на дерево? Это же тот же самый метод получается...
Сделайте его обычным методом.

YaroslavKomarov replied:
Мне почеу-то показалось, что это имеет смысл. исправлю

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.h:
```diff
@@ -0,0 +1,26 @@
+
+class MTreeNode
+{
+private:
```
Уже писал.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+
+MTreeNode* MTreeNode::begintTree(int i, int j) {
+	MTreeNode* node = new MTreeNode(nullptr);
+	node->m_distance = 0;
```
Этого можно не делать, если он при объявлении инициализирован будет...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+
+MTreeNode::MTreeNode(MTreeNode* parent) {
+	m_parent = parent;
+	m_distance = 0;
```
Ой вэй! Таки при живом родителе дистанс 0???

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+	m_parent = parent;
+	m_distance = 0;
+	m_i = 0;
+	m_j = 0;
```
Лишнее...

YaroslavKomarov replied:
Забавно, несмотря на инициализацию при объявлении VS сам подчеркивает "желтеньким", что не все поля инициализированы, но исправил.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+
+MTreeNode::~MTreeNode() {
+	for (auto& ptr : m_children)
```
Тут можно без ссылки, это разумно. Ибо ссылка столько же места занимает, сколько и указатель, от неё только хуже -- лишние прыжки по указателям.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+
+int MTreeNode::i() const { return m_i; }
+
+int MTreeNode::j() const { return m_j; }
```
Мама мия. Уберите в заголовочник все однострочники.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+bool MTreeNode::addChild(int i, int j) {
+	if (abs(i - m_i + j - m_j) != 1)
+		return false;
+	m_children.push_back(new MTreeNode(this));
```
БРРРР... А если такой уже есть??? Давайте ка проверять сначачла!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+
+MTreeNode* MTreeNode::hasChild(int i, int j) {
+	int count = m_children.size();
```
const уж тогда.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+
+MTreeNode* MTreeNode::hasChild(int i, int j) {
+	int count = m_children.size();
+	for (int ind = 0; ind < count; ++ind) {
```
А вообще, пишите тогда правильный for(auto child : m_children)

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+	queue<MTreeNode*> nodes;
+	MTreeNode* parent = tree.m_parent;
+	if (parent) {
+		if (parent->i() == i && parent->j() == j)
```
Зачем этот страх сделан?
Ищите вниз по дереву от того узла, который вам дали.
Либо напишите отдельно проход до корня и ищите от него, если вы хотите, чтобы вам было без разницы, какой узел дерева отдавать для поиска.
Переделать. Ужас!

YaroslavKomarov replied:
Могу я в таком случае изменить сигнатуру метода searchNode, убрав вообще передачу const ссылки на узел? Отдельным методом буду проходить от объекта, из которго был вызван метод, до корня и потом уже искать узел во всем дереве

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,91 @@
+		}
+	}
+	return nullptr;
+}
```
Пустые строкииии....

Toxin65 replied:
Так и не исправили...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+	UpDownRight = Up + Down + Right,
+	LeftDownRight = Left + Down + Right,
+	AllDirection = Up + Left + Down + Right,
+};
```
"Наркота то какая.... Наркотаааааа....."
Ну я сам напросился. В целом грамотно, битовые флаги, все дела, енум. Всё как учили.
По хорошему, под это дело бы отдельный класс сделать. Но мы учимся, так что норм.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.h:
```diff
@@ -0,0 +1,20 @@
+
+class Maze {
+private:
+	map<int, char> m_dict;
```
Как говорит мой коллега: "Не надо делать это членом!"
Уберите в *.cpp и сделайте его там статиком. И инклуд отсюда уберите.

YaroslavKomarov replied:
Ваш коллега очень прав). Я сам немного удивлен тому, что написал. Исправлю

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.h:
```diff
@@ -0,0 +1,20 @@
+	map<int, char> m_dict;
+	MCell* m_field;
+	int width;
+	int size;
```
инит всё и вся
А ещё, я не сразу заметил, храните ширину/высоту, а не вот этот ужас!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.h:
```diff
@@ -0,0 +1,20 @@
+	Maze(int n, int m);
+	~Maze();
+	const MCell& cell(int i, int j) const;
+	bool hasConnection(int i1, int j1, int i2, int j2);
```
const

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+Maze::Maze(int n, int m) {
+	width = m;
+	size = n * m;
+	m_field = new MCell[size];
```
То m_, то не m_...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+	m_dict = { {AllDirection, char(197)}, {UpLeftDown, char(180)}, {UpLeftRight, char(193)},
+		{UpDownRight, char(195)}, {LeftDownRight, char(194)}, {UpLeft, char(217)},
+		{UpDown, char(179)}, {UpRight, char(192)}, {LeftDown, char(191)},
+		{LeftRight, char(196)}, {RightDown, char(218)}, {Up, '0'}, {Down, '0'}, {Left, '0'}, {Right, '0'} };
```
Вот вот! Статиком его! Статиком!
Он вот нафиг не нужен каждому объекту.
Дак ещё и const его, ибо нефиг его менять!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+
+Maze::~Maze() { delete[] m_field; }
+
+const MCell& Maze::cell(int i, int j) const { return m_field[i * width + j]; }
```
Надо ассертиться, если спрашивает чего не положено!
А то отдадите объект, которого нет, т.е. чужую память!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+
+bool Maze::hasConnection(int i1, int j1, int i2, int j2) {
+	int ind1 = min(i1, i2);
+	int ind2 = min(j1, j2);
```
const оба

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+	if (j2 < 0 || j2 >= width || i2 < 0 || i2 >= size - width)
+		return false;
+	if (i1 == i2) {
+		return m_field[ind1 * width + ind2].right();
```
Элегантно!

YaroslavKomarov replied:
Пожалуйста, скажите, если тут что-то требуется поправить. А если Вы искренне, то спасибо!)

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+bool Maze::hasConnection(int i1, int j1, int i2, int j2) {
+	int ind1 = min(i1, i2);
+	int ind2 = min(j1, j2);
+	if (j2 < 0 || j2 >= width || i2 < 0 || i2 >= size - width)
```
Надо проверять, что они соседи по вертикали или по горизонтали!
А то дадут вам диагональ, или вообще не соседей!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+	}
+	else
+		return false;
+}
```
Тут при неправильных индексах надо бы ассерт сделать. Хотя и не обязательно.
Ну и замечания выше.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+	}
+	else if (j2 == j1) {
+		m_field[ind1 * width + ind2].m_down = false;
+		return true;
```
Если связи не было -- тоже false.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+		int j = n % width;
+		int temp1 = m_field[n].down() << 2;
+		int temp2 = m_field[n].right() << 3;
+		int temp3 = temp2 | temp1;
```
Этот вообще не используется!

YaroslavKomarov replied:
Ой! это я при отладке использовал, забыл удалить.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+			res += m_field[i * width + j - 1].right() << 1;
+
+		if (res == 0 || res == 1 || res == 2 || res == 4 || res == 8)
+			cout << '0';
```
Дак уже добавить его в m_dict...
И не нужен этот глупый if...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,108 @@
+		if (res == 0 || res == 1 || res == 2 || res == 4 || res == 8)
+			cout << '0';
+		else
+			cout << m_dict.find(res)->second;
```
Такс такс. А если не найдёте? Ну вдруг?
Это же будет пустой итератор!!! Атата!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+using namespace std;
+
+void buildFullMaze(Maze& iMaze, const MTreeNode* tree);
+void BuildMaze2(Maze& maze, const int& maze_size);
```
const int&
Но зачем по ссылке то??? Он же не меняется в процессе выполнения функции.
Внимательно изучите когда ссылка нужна. Тут -- не нужна точно. И ниже тоже.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+void buildFullMaze(Maze& iMaze, const MTreeNode* tree);
+void BuildMaze2(Maze& maze, const int& maze_size);
+void BuildMaze1(Maze& maze, const int& maze_size);
+void GetTree(MTreeNode* node, const int& width, const int& height);
```
Написано Get, ничего не возвращает. Build чтоли?

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+	int sum_weights = 0;
+	int max_weight = 0;
+	queue<const MTreeNode*> nodes;
+	vector<vector<string>> to_print(height);
```
Вектор векторов -- это ужас. Я даже на вскидку не придумаю, когда это действительно нужно.

YaroslavKomarov replied:
Я просто принадлежу к поколению наглых питонистов, которые творят, что хотят))) Но если серьезно, то все-таки стоит это переделать или можно оставить?

Toxin65 replied:
Да пусть будет...
Но предупреждение вы получили.
При приёме на работу я такую фигню не потерплю!!!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+void PrintTreeInfo(const MTreeNode* tree, const int& width, const int& height);
+vector<int> GetRandCoordinates(const int& width, const int& height);
+
+void PrintTreeInfo(const MTreeNode* tree, const int& width, const int& height) {
```
Ох, как плохо без комментариев.
Что делает эта функция? Зачем программист её написал?

YaroslavKomarov replied:
Функция для получения произвольных координат на границе рандомного лабиринта. Теперь этот комментарий прописал в коде

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+	vector<vector<string>> to_print(height);
+	for (auto& vctr : to_print)
+		vctr = vector<string>(width);
+	while (true) {
```
for(;;)

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+		int i = tree->i();
+		int j = tree->j();
+		sum_weights += tree->distance();
+		max_weight = tree->distance() > max_weight ? tree->distance() : max_weight;
```
std::max

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+	}
+	for (auto& vctr : to_print) {
+		for (auto& str : vctr)
+			cout << setw(5) << str;
```
Неплохо...

YaroslavKomarov replied:
Тоже забыл это поправить перед отправкой! В setw() теперь передаю длину строки, полученной от вершины с максимальным весом + 1 (const int max_len = size(to_string(max_weight) ....setw(max_len) << str...)

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+			cout << setw(5) << str;
+		cout << endl;
+	}
+	int average_weight = sum_weights / (1. * width * height);
```
const

Toxin65 reviewed (changes requested):
Исправьте то, что есть. Потом продолжим.
Извините, немного недочитал, надо уделить внимание всем, по возможности...

YaroslavKomarov committed:
<F12#

YaroslavKomarov committed:
Все поля проинициализированны при объявлении;
Однострочные методы инициализированны в заголовочных файлах;
Убрал из заголовочников

YaroslavKomarov committed:
Все поля проинициализированны при объявлении;
Одностроные методы проинициализированны в заголовочных файлах;
В заголовочных больше не используются пространства имен;
const есть везде, где можно;
Исправлен нелепый метод searchNode;
Добавлен метод поиска корня в дереве searchRoot;
во всех заголовочниках сначала идет public-секция;
При живом родителе у дочернего элемента distance правильный;
Лишние ссылки при прохождении по векторам (и при int) убраны;
Перед добавлением ребенка происходит проверка на его наличие к родителя;
m_dict из класса Maze больще не член класса и он sctatic;
добавлены assert при попытке получения клетки лабиринта и где-то еще;
Добавлена проверка на соседство двух клеток для методов класса Maze;
Теперь, если звязи не было метод removeConnection возвращает корректное
значение;
Вообще, кажется, все исправил, что было прописано Вами.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MCell.cpp:
```diff
@@ -0,0 +1,5 @@
+//#include "MCell.h"
```
Это можно было бы и остаивть. Тогда он откомпилится в отдельный объектный файл. А это неплохо!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.h:
```diff
@@ -0,0 +1,26 @@
+	const MTreeNode* searchNode(const int i, const int j);
+	const MTreeNode* searchRoot(const MTreeNode* node);
+private:
+	int m_i, m_j = 0;
```
Что-то тут копипастой запахло.
Где-то уже видел...

YaroslavKomarov replied:
Вы сказали, что бы я исправил былое безобразие. Вы мне вариантом предложили сделать отдельный метод для поиска корня, если я хочу не зависеть от передаваемого узла. Я так и поступил.

Toxin65 replied:
Дык не про то речь... Ну и ладно...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+
+MTreeNode::MTreeNode(MTreeNode* parent) {
+	m_parent = parent;
+	m_distance = parent == nullptr ? 0 : parent->m_distance + 1;
```
Можо не тернарным, а просто при наличии родителя у него спросить и +1...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+		if (child->m_i == i && child->m_j == j)
+			return false;
+	m_children.push_back(new MTreeNode(this));
+	m_children[m_children.size() - 1]->m_distance = this->m_distance + 1;
```
Просто используйте back...
И ещё, m_distance вы в конструкторе считаете. Зачем это здесь тогда?

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+}
+
+MTreeNode* MTreeNode::hasChild(int i, int j) const{
+	const int count = m_children.size();
```
Это тут зачем. Осталось?

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+
+MTreeNode* MTreeNode::hasChild(int i, int j) const{
+	const int count = m_children.size();
+	for (auto child : m_children){
```
Разумно не делать копии, то есть auto&
Хотя при нашем размере, можно и копию себе позволить...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+	return nullptr;
+}
+
+const MTreeNode* MTreeNode::searchNode(const int i, const int j) {
```
Документировать было бы неплохо! Ибо вы корень дерева ищите, а не от текущей ноды ищите!
Это поведение неочевидно окружаюцим.
Хороший код не нуждается в комментариях.
Хорошие методы прокомментированы так, чтобы не читать их код.

Toxin65 replied:
Доки надо в заголовках. Ну да ладно... На будущее...
Ну и кодировка похерилась...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+				return child;
+			nodes.push(child);
+		}
+		if (nodes.size() == 0)
```
Это можно вынести в while, так то!

YaroslavKomarov replied:
Я могу заблуждаться, но ведь есть ситуация при которой мы достанем новый узел из очереди и она опустеет, однако этот узел будет иметь детей, которых потом также потребуется обойти. Мы прервем цикл до их обхода.

Toxin65 replied:
Ошибаетесь, подумайте...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+const MTreeNode* MTreeNode::searchRoot(const MTreeNode* node) {
+	queue<MTreeNode*> nodes;
+	MTreeNode* parent = node->m_parent;
+	if (parent) {
```
Следует уменьшать вложенность. if(!parent) return node; Конечно же с переносами..
Это делает код читабильнее.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,115 @@
+		{UpDownRight, char(195)}, {LeftDownRight, char(194)}, {UpLeft, char(217)},
+		{UpDown, char(179)}, {UpRight, char(192)}, {LeftDown, char(191)},
+		{LeftRight, char(196)}, {RightDown, char(218)}, {Up, '0'}, {Down, '0'}, {Left, '0'}, {Right, '0'}, {0, '0'}
+};
```
Молодца! статик!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,115 @@
+
+const MCell& Maze::cell(int i, int j) const { 
+	assert(i >= 0);
+	assert(j >= 0);
```
Надо бы на рамерность тоже проверять. А то в чужую память залезем как нефиг делать... А это очень плохо!
И пишите одним ассертом. Так лучше. Можно разделять, когда они в разных местах...

Toxin65 replied:
Вот это критично очень!!!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,115 @@
+bool Maze::hasConnection(int i1, int j1, int i2, int j2) const {
+	if ((abs(i1 - i2) + abs(j1 - j2) != 1)) return false;
+	if (j2 < 0 || j2 >= m_width || i2 < 0 || i2 >= m_height)
+		return false;
```
Может вынести эти штучки отдельно. А то дублирование получается. Ниже то же самое написано...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,115 @@
+		return m_field[ind1 * m_width + ind2].down();
+	}
+	else
+		return false;
```
Тут вообще ассерт, такого быть то не должно в природе!
Ассерт -- хорошая проверка программиста, что он что-то накосячил...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,115 @@
+	if (j2 < 0 || j2 >= m_width || i2 < 0 || i2 >= m_height)
+		return false;
+	int ind1 = min(i1, i2);
+	int ind2 = min(j1, j2);
```
const им всем же!

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Maze.cpp:
```diff
@@ -0,0 +1,115 @@
+		if (j + 1 == m_width)
+			cout << endl;
+	}
+}
```
Пустая строка же!!!

Toxin65 reviewed (changes requested):
Давайте поправим...
Всё бы хорошо. Но есть непростительное!

YaroslavKomarov committed:
Убрал MCell.cpp;
в конструкторе MTreeNode упразднил тернарный оператор;
заменил индексацию для вектора в методе addChild на .back;
убрал вычисление m_distance из addChild;
убрал лишнее вычисление длинные вектора детей в hasChild;
Вернул ссылку для итерирования по вектору в hasChild;
Прописал несколько комментариев для serchNode, searchRoot;
Убрал return nullptr у метода searchNode;
Уменшил вложенность для большей читаемости кода у searchRoot;
Исправил assert у метода cell класса Maze;
Вынес в отделный метод checkConnection проверку на вшивость для всех связей;
Добавил в метод checkConnecion проверку нереального случая;
добавли нужные const;
довил пустую строку в конце Maze.cpp.

YaroslavKomarov committed:
Добавил пустую строку в конце каждого файла.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+#include <iostream>
+#include <queue>
+#include "MTreeNode.h"
+using namespace std;
```
Стоит отделять пустой строкой разные вещи для читабельности...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+
+MTreeNode::~MTreeNode() {
+	for (auto ptr : m_children)
+		delete ptr;
```
У вас проблемы с удалением...

YaroslavKomarov replied:
Речь о том, что я мог просто вызвать delete[] ? Или я оставляю в памяти висеть детей того родителя, у которого вызывается деструктор?

Toxin65 replied:
Сорян, всё норм. Я уже обревьювился в конец...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+using namespace std;
+
+MTreeNode* MTreeNode::begintTree(int i, int j) {
+	MTreeNode* node = new MTreeNode(nullptr);
```
Что тут происходит? Один элемент выделили? Кошмар! Это неправильно, видимо. Да и зачем это тут? Есть несколько вариантов, как себя вести. Это поведение точно ошибочное!

YaroslavKomarov replied:
Тут утечка, верно? Кстати говоря, я вызывал delete для корня дерева в practise_18.11.20.cpp в случае построения второго лабиринта (про третий благополучно забыл). Но я догадываюсь, что это не есть хорошо. Можете намекнуть на пример корректного поведения для beginTree?

Toxin65 replied:
Извиняюсь, запутался... Всё норм.

Toxin65 reviewed (commented):
Беды с памятью... Большие беды! Разберитесь с new delete..

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.h:
```diff
@@ -0,0 +1,25 @@
+	const MTreeNode* child(int i) const { return m_children[i]; }
+	static MTreeNode* begintTree(int i, int j);
+	const MTreeNode* searchNode(const int i, const int j);
+	const MTreeNode* searchRoot(const MTreeNode* node);
```
Оба метода сами по себе const.

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+#include "MTreeNode.h"
+using namespace std;
+
+MTreeNode* MTreeNode::begintTree(int i, int j) {
```
Ну ё-маё... А я его ищу. beginTree!!! А не begint...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/MTreeNode.cpp:
```diff
@@ -0,0 +1,78 @@
+}
+
+MTreeNode* MTreeNode::hasChild(int i, int j) const{
+	for (auto &child : m_children){
```
Нет смысла ссылку делать. Получается ссылка на указатель -- это тут лишнее...

Toxin65 commented on Practise_18.11.20/Practise_18.11.20/Practise_18.11.20.cpp:
```diff
@@ -0,0 +1,210 @@
+int main()
+{
+	const int maze_size = 5;
+	Maze maze1 = Maze(maze_size, maze_size);
```
`Maze maze1(maze_size, maze_size);`

Toxin65 reviewed (approved):
Не идеально... Но зачесть можно.

YaroslavKomarov committed:
Добавил const к методам searchNode, searchRoot;
Конечно же исправил beginTree)));
убрал ссылку в hasChild;
Теперь корректно создаю экземпляры класса Maze.
